### DISCUSSION TITLE
Reduce freezes due to RPC calls

### DISCUSSION THREAD

**Discussion by raisjn** [nikitabobko/AeroSpace#655] [feature-ideas] [answered]
**Running into slow Aerospace?** 

EDIT: 12/05: ported perf patch to latest origin/main branch

Try out [this branch](https://github.com/raisjn/AeroSpace/tree/dev/okay/2024_12_05_less_cpu_lockups). 

**Usage** 

Recompile the debug or release build and add the following lines to your aerospace.toml:

```
new-window-detection-timeout = 100
```

@liamaharon suggests the following may have better results:

```
new-window-detection-timeout = 50
new-window-detection-debounce = 100
```

What it does: `new-window-detection-timeout` is how long to block the main thread (up to X ms), while `new-window-detection-debounce` will coalesce calls within X ms together. 

**How it works**

The root cause of slowness is that Aerospace iterates over windows and uses accessibility APIs to communicate with each app. If any application is slow to respond, it blocks Aerospace's main thread from executing. To prevent Aerospace from hanging, we put the communication with apps into a new thread.

**Details**

`detectNewWindowsAndAttachThemToWorkspaces` calls into a debouncer and a separate queue when called from refreshSession and wait for up to X ms for it to complete. `validateStillPopups` is behind a similar debouncer + dispatch queue. 

*Note*: Thread safety: for detectNewWindowsAndAttachThemToWorkspace to be threadsafe, MacWindow.get() has an NSLock placed in it so that multiple threads can access it safely.   Additionally, specific work within MacWindow.get() should be moved to the main thread using a DispatchQueue

**Observed Bugs:**

Bug A: Starting a popup window and dragging it across monitors doesn't work (for example, System Settings). The window will get stuck on a single monitor and keep hopping back.  Dragging the window within a single monitor works fine. It is possible to move the window via keyboard shortcuts, though. (I think this happens on all approaches? I haven't tested stock Aerospace)

Bug B: Focus follows mouse (AutoRaise application) doesn't play well with aerospace and multiple monitors. Switching the workspace on one monitor (monitor 1), then moving to another monitor (monitor 2) with the mouse will often cause the workspace on monitor 1 to switch back to the previous workspace. (I think this happens on all approaches? I haven't tested stock Aerospace)

**FAQ**

1. When will this be merged upstream? Answer: I have no plan to attempt to upstream this. I keep a private fork with my own changes.  I've released this particular patch because [people have been asking for the last year for performance improvements](https://github.com/nikitabobko/AeroSpace/issues/131). 
2. How do I install this? Answer: See @wojciech-kulik's [comment](https://github.com/nikitabobko/AeroSpace/discussions/655#discussioncomment-11460077)
3. Did Aerospace crash because of this patch? Answer: Maybe. I believe [Aerospace itself is prone to crashing](https://github.com/nikitabobko/AeroSpace/issues?q=is%3Aissue+crash).

**Comment by raisjn** [comment]
> Thanks to @raisjn for doing the performance experiments. The debouncing approach makes sense. The refreshSession may be called too often when a lot of things happen on the screen (e.g. when you switch the workspace), and given the idempotent nature of refreshSession, debouncing the whole session becomes trivial and may increase the app responsiveness. I don't agree with the specifics of implementation, but the general idea makes sense, thanks for your experiments. I will consider "the debouncer" approach when start working on performance.

@nikitabobko: note my comment about debouncing refreshSession having some small bugs in it. if you can get it working, that is great! overall, I don't think solely debouncing refreshSession will be enough to stop hangs: if any app misbehaves, you will eat the 1s timeout (per app) and that is already too much delay. 

PS: iiuc, refreshSession is idempotent if no body is passed in, but when a body is passed in, then it is executing code that has side effects. 

> As correctly pointed by @raisjn, please discuss the details of @raisjn fix in https://github.com/nikitabobko/AeroSpace/discussions/655

if you can re-parent comments from #131 here, that would be great

**Reply by nikitabobko** [reply]
Sorry for the late response, I was on a vacation

> note my comment about debouncing refreshSession having some small bugs in it

If you describe what bugs you observed, maybe I can quickly tell you what was wrong. But I think you will rather have more bugs (especially concurrent bugs) if you try to debounce only the part of the `refreshSession`

> PS: iiuc, refreshSession is idempotent if no body is passed in, but when a body is passed in, then it is executing code that has side effects.

Yes

> I don't think solely debouncing refreshSession will be enough to stop hangs: if any app misbehaves, you will eat the 1s timeout (per app) and that is already too much delay.

Yes, https://github.com/nikitabobko/AeroSpace/issues/131#issuecomment-2302135213

> if you can re-parent comments from https://github.com/nikitabobko/AeroSpace/issues/131 here, that would be great

GitHub doesn't allow it it. It's only possible to convert the whole issue to the discussion

**Comment by liamaharon** [comment]
fyi @raisjn (and anyone else who may be experiencing this issue)

```
new-window-detection-timeout = 50
new-window-detection-debounce = 100
```

on `dev/okay/2024_11_02_less_cpu_lockups` is working wonderfully for me in initial experiments. Much appreciated ❤️

**Comment by vianneychin** [comment]
I'd love to try your build! Stupid question, but what would the step by step look like to install the forked version? I normally get it from homebrew.

**Reply by wojciech-kulik** [reply]
Download repo, run `bundle install`, run `./install-from-sources.sh`. I had to remove also from `build-shell-completions.sh` the following line and commit changes locally:
 ```
fish -c 'source ./.shell-completion/fish/aerospace.fish'
```

Update:
If you encounter some issues with the command, you may need to create a signing certificate: https://github.com/nikitabobko/AeroSpace/blob/main/dev-docs/development.md#2-create-codesign-certificate

and call the command with an extra `PATH`:
```
PATH="/path/to/repo/.deps/cargo-root/bin:$PATH" ./install-from-sources.sh
```

**Reply by quarkquartet** [reply]
Sorry for a silly question. I just successfully run `./install-from-sources.sh`. But where is the installed app?

**Reply by wojciech-kulik** [reply]
It should be in `/Applications` :)

**Reply by quarkquartet** [reply]
> It should be in `/Applications` :)

<img width="1258" alt="image" src="https://github.com/user-attachments/assets/c62cb304-9bd5-4647-8229-efae47ae1cca" />

This is what I get from `./install-from-sources.sh`, but still did not see in the /Applications

**Reply by wojciech-kulik** [reply]
The warning tells what's wrong. Try:
```
PATH="/path/to/repo/.deps/cargo-root/bin:$PATH" ./install-from-sources.sh
```

**Reply by quarkquartet** [reply]
> The warning tells what's wrong. Try:
> 
> ```
> PATH="/path/to/repo/.deps/cargo-root/bin:$PATH" ./install-from-sources.sh
> ```

Thank you! That almost solved the problem. I think I got stuck now in the last step. Sorry I'm not familiar with how an Mac App is build.
<img width="1258" alt="image" src="https://github.com/user-attachments/assets/7ee62fac-067d-46c1-b561-c59482c361d6" />

**Reply by wojciech-kulik** [reply]
You should have `aerospace-codesign-certificate` in your Keychain. I'm not sure why it's missing on your side.

**Reply by quarkquartet** [reply]
> You should have `aerospace-codesign-certificate` in your Keychain. I'm not sure why it's missing on your side.

Thank you for reminder! I searched and found the instruction about creating this keychain. Now I successfully installed it!

**Reply by wojciech-kulik** [reply]
Here is the instruction: https://github.com/nikitabobko/AeroSpace/blob/main/dev-docs/development.md#2-create-codesign-certificate if anyone else faces the same issue.

**Comment by PhotonQuantum** [comment]
Thank you for your excellent work! While the original version is slow to the point of being unusable, `dev/okay/2024_12_05_less_cpu_lockups` works like a charm on my setup! Is there any plan to get this merged upstream?

**Reply by voidpointer0** [reply]
Would love to know too, hoping this will help with the system becoming unusable after launching Unity.

**Reply by quarkquartet** [reply]
Yeah this seems to be a nice solution. Currently it's really lagging as long as I'm running some code on my laptop.

**Reply by PhotonQuantum** [reply]
Followup: so today I got a runtime error on this version
https://gist.github.com/PhotonQuantum/801a680257ceded3d3ff5fc662557fe2

I wasn't doing anything particular when it crashed. The lid was closed at that time. I have an external monitor setup that could be related to disconnection. Still, I can't be sure because I can't find any relevant event in the system log at that exact timestamp.

@raisjn I'm uncertain if your fork introduces the issue or if it is a general problem from upstream, so I'm posting it here for now.

**Comment by quarkquartet** [comment]
@raisjn Thank you for your solution. Just make sure I understand it correctly: the version installed from Homebrew does not have your implementation, and I have to install it from your forked repo with `dev/okay/xxx_less_cpu_lockups` right?

**Comment by grapexy** [comment]
Just wanted to report that I've been running `2024_12_05_less_cpu_lockups` for about 3 weeks now without any issues. I was having issues with upstream version of aerospace locking up when running some local LLMs, and all these issues are now gone.

**Reply by vianneychin** [reply]
The performance is just really good as someone who went through the trouble of compiling a forked version. Spent literally 2 hours figuring it out but was worth it lol

**Comment by voidpointer0** [comment]
Any estimate on when/if this will be merged to main? I could really use it, but I don't want to mess with my current installation.

**Comment by wojciech-kulik** [comment]
@raisjn would it be possible to update your branch to include latest changes from the main repo?

**Reply by raisjn** [reply]
i rebased and pushed [2025_01_12_less_cpu_lockups](https://github.com/raisjn/AeroSpace/tree/dev/okay/2025_01_12_less_cpu_lockups), let me know how it works. i will try it out this week as well

is there any particular reason you wanted the latest changes from main?

**Reply by wojciech-kulik** [reply]
I just wanted to make sure that all the fixes are included, especially that I saw some crash fixes :). Thank you very much!

**Reply by wojciech-kulik** [reply]
I just got this crash :(

```
##### AeroSpace Runtime Error #####

Please report to:
    https://github.com/nikitabobko/AeroSpace/discussions/categories/potential-bugs
    Please describe what you did to trigger this error

Message: 
Version: 0.0.0-SNAPSHOT
Git hash: 789435f1e65c4986abbc51e90e662b252f57b94a
Date: 2025-01-13 21:57:20 +0000
Coordinate: AppBundle/MacWindow.swift:354:10 onWindowDetected(_:startup:)
recursionDetectorDuringFailure: false
cli: false
refreshSessionEvent: nil
Displays have separate spaces: true

Stacktrace:
0   AeroSpace                           0x000000010243f6f0 $s6Common6errorT_4file4line6column8functionxSS_SSS2iSStlFs5NeverO_Tg5 + 1388
1   AeroSpace                           0x000000010246b170 $s6Common5check__4file4line6column8functionySb_SSyXKSSS2iSStF + 92
2   AeroSpace                           0x0000000102304160 $s9AppBundle16onWindowDetected33_AFFA185B3FDC2EA3EF5AFA96A7308179LL_7startupyAA0D0C_SbtF + 428
3   AeroSpace                           0x00000001022f3d58 $s9AppBundle19validateStillPopups33_EA737387E8EE73FEB20FFAD7FE91D51ELL7startupySb_tF + 276
4   AeroSpace                           0x000000010231e024 $s9AppBundle15TimeoutDetectorC3run4task8onFinishyyACc_yACctFyyYbcfU_ + 44
5   AeroSpace                           0x000000010231dcf0 $sIeg_IeyB_TRTm + 28
6   libdispatch.dylib                   0x00000001808f28f8 _dispatch_call_block_and_release + 32
7   libdispatch.dylib                   0x00000001808f4658 _dispatch_client_callout + 20
8   libdispatch.dylib                   0x00000001809064e8 _dispatch_root_queue_drain + 860
9   libdispatch.dylib                   0x0000000180906b20 _dispatch_worker_thread2 + 156
10  libsystem_pthread.dylib             0x0000000180aa339c _pthread_wqthread + 228
11  libsystem_pthread.dylib             0x0000000180aa20f0 start_wqthread + 8
```

**Reply by raisjn** [reply]
after testing for the last week, i've found it to be unstable - likely a combination of upstream changes and not properly guarding against them. i've moved back to using the previous version

**Reply by wojciech-kulik** [reply]
I wonder if we ever get to the point where Aerospace works well without "3rd-party" hacks :(... It seems like the developer is not really interested in fixing this problem. The only reason I use Aerospace is to switch between workspaces instantly and that feature is not working well. 

Maybe it would be a good idea to create another open-source project that manages virtual workspaces without any windows tiling and we could focus on making it blazingly fast instead of this laggy behavior.

**Comment by wojciech-kulik** [comment]
📢 It doesn’t seem like it’s ever going to be fixed, so I decided to start working on my own virtual workspace manager focused only on switching workspaces without tiling. After a few hours, I finished an MVP, which you can check out here:

<img width="781" alt="FlashSpace" src="https://github.com/user-attachments/assets/a8633971-1473-4b70-9998-bb176185ebb9" />

https://github.com/wojciech-kulik/FlashSpace

- This solution seems to work fine even with high CPU usage
- It's probably buggy, I've tested it only on my own setup
- Even if my approach is not sufficient, now we've got a solid base to experiment and find the best solution. There is only one place that handles all the logic of switching workspaces.

**Reply by nikitabobko** [reply]
> It doesn’t seem like it’s ever going to be fixed

> It seems like the developer is not really interested in fixing this problem

@wojciech-kulik The way you talk is rude and demotivating (not necessarily towards me. For example, this is also not a good way to communicate: https://github.com/koekeishiya/yabai/issues/2010#issuecomment-1863599401).

You are welcome to talk about your new shiny project, but you don't have to be mean about others people work, focus on your project instead.

One can understand that this performance problem is the priority for me by looking at the pinned issues.

You have received the first and the only warning about your manner of communication.

**Reply by wojciech-kulik** [reply]
Well, I'm not even going to comment on you pulling some of my random comments, because it's quite childish behavior and I'm not going to post here elaborate explaining the context. 

Regarding the situation here, to give some context, I contributed to this repository, I promoted this project before anybody knew what Aerospace is, I suggested to Josean Martinez to make a YouTube video about your project, and I promoted it on [reddit](https://www.reddit.com/r/MacOS/comments/18w0aej/aerospace_probably_the_best_window_manager_for/) as well, so I believe I have some right to have an opinion on the way this project is managed, even though it's a tiny right. (btw. interesting why I don't have a contributor badge here - is it because you close pull requests instead of merging/squashing them - to avoid people being tagged as contributors?)

You, on the other hand, ignore crucial performance issues forcing people to use some 3rd party forks not even giving a clear message or plan for fixing this problem. This discussion has been already open for 3 months. To emphasize the importance of this issue, one person even donated you $500 to fix it and the only thing you did was posting a message justifying your decisions and locking the issue from further comments without any clear plan. Not a very open-source-friendly approach.

So yes, I believe I've got a right to say: "It seems like the developer is not really interested in fixing this problem", "It doesn’t seem like it’s ever going to be fixed". Sorry if it offends you, but I would feel more offended by supporting this project that much and being ignored in such a crucial bug (I'm not talking just about me). Many people stopped using this project because of this issue, even Josean Martinez who made a video about your project. So yes, it is an extremely important bug that is completely ignored by you.

> One can understand that this performance problem is a priority for me by looking at the pinned issues

Pinning the issue is not a sign of something being important to you. More like avoiding more discussions in this area. This issue has been open for 3 months (and the parent issues for 1 year+) and this is your first comment here, truly important thing to you... This is my last and only message to you here.

**Reply by saivan** [reply]
@nikitabobko I just thought I'd chime in here. I've been pretty patient with this bug because I loved what you'd done with Aerospace and honestly this is a great piece of software. So please understand that what I say here, I say because I really want to see you succeed. I don't want it to come off as an attack or a criticism, I want you to be great because what you've produced is so impressive!

I'm jumping ship today, which really makes me sad because I've really loved this software. But this bug has slowed me down so much that it sometimes feels like I'm using a computer from 2010. For every one person that is very vocal about a bug, there are going to be hundreds that just stay quiet and watch how you respond.

You've had a lot of signals that this bug is making your software unusable for a lot of people and it's been left unfixed for a very long time. It's fine to tell everybody "hey this isn't my priority, I've got other things going on with my life", which is a completely fair response and I don't believe too many people would be upset with you. 

I wish you the best with Aerospace, but unfortunately - this is something that is going to have you losing users at a rapid rate. Please make it a priority to fix it if you truly want people to use what you build. If that isn't of importance to you, then all I can say is - thank you for everything. I've truly enjoyed using aerospace while it was my daily driver for the last eight months. Thank you for all of your hard work.

**Reply by grapexy** [reply]
Would also like to chime in here. I can't abandon AeroSpace because I honestly can't imagine living without it, however, I have skipped all the AeroSpace updates and stayed on the hotfix branch since @raisjn dropped it and will remain on it until it stops working for some reason.

I don't really need any new updates, so I'm fine on staying on this hotfix branch, but at the same time, I can't recommend AeroSpace to anyone because this fix is not in upstream and installing a fork is quite tedious for non-developers.

What should really be taken into consideration is that a lot of people are working with various local LLMs and they are very resource heavy. I've had AeroSpace frozen for more than 30 seconds in upstream version during one of the heavier usage times, on a top of the line MacBook.

**Reply by saivan** [reply]
Actually. In the last two hours. I have found a solution that really does work quite well as a replacement for Aerospace. Unfortunately, I don't get tiling, which I will sorely miss. But this solution is almost completely native and won't slow down your machine very much (if at all):

1. Make as many desktops as you need spaces (I have E for editor, W for web, S for spotify, etc...)
2. (optional) Remap your caps lock key to a modifier, there are three options here:
  - Native: Remap your caps lock key to option (go to System Settings > Keyboard > Keyboard Shortcuts > Modifier Keys, and change the Caps Lock key's action to "Option") 
  - or use something like Karabiner elements to remap caps lock to multiple keys (I've included my script to the bottom)
  - or use the built in hyper key in ray-cast
3. Go into your keyboard settings and set native keyboard shortcuts to go to each space 
4. You can just use a simple window management tool like betterSnapTool to make hot zones that you can snap to

<img width="827" alt="image" src="https://github.com/user-attachments/assets/6f012d35-e173-45b2-9210-a3ee15cbd394" />

Again @nikitabobko I'd love to come back to aerospace someday, but for those of use that are finding it unserviceable at the moment, this should work for now.

## Karabiner Config

If you choose to use carabiner, this config will create a hyper key for you. Raycast can also do this natively if you like

```
{
    "description": "Caps Lock to Esc when pressed alone, and Hyper key (Command+Control+Option+Shift) when held",
    "manipulators": [
        {
            "from": {
                "key_code": "caps_lock",
                "modifiers": { "optional": ["any"] }
            },
            "to": [
                {
                    "key_code": "left_shift",
                    "modifiers": ["left_command", "left_control", "left_option"]
                }
            ],
            "to_if_alone": [{ "key_code": "escape" }],
            "type": "basic"
        }
    ]
}
```

**Comment by acsandmann** [comment]
This workaround used in yabai may show some promise in resolving these issues in a relatively simple manner...

https://github.com/koekeishiya/yabai/commit/0e3e999de1b51af0aa1eccd379b857402aa39d68

**Comment by voidpointer0** [comment]
Is someone working on fixing those lags introduced by Unity etc.? I reached a point where I have to switch to something else because it makes my work impossible.

**Comment by rickyz** [comment]
I did some analysis of the sources of Aerospace latency and came to the same conclusions described here - specifically:

On operations like workspace changes, aerospace receives a stampede of accessibility events, each of which triggers a serial call to `refreshSession()` on the main thread. Since these `refreshSession()` calls are idempotent, we should be able to coalesce these calls so that multiple accessibility events can be handled by a single `refreshSession()` call. While this by itself doesn't fully address the issue with hanging accessibility queries described here, it's a pretty straightforward change that noticeably improves latency for me: https://github.com/rickyz/AeroSpace/commit/a973bc4158834c4887712bdd82ba42b99b76c56c

With this change, operations like workspace switch go from O(num_source_windows + num_dest_windows) to O(1) `refreshSession()` calls (with similar improvements for operations like closing windows or resizing windows).

If this change seems reasonable, I'd be happy to send a PR and address any comments!

**Reply by nikitabobko** [reply]
The negative consequence of your approach is that it cancels **the new** request, ideally, it should cancel **the currently running** request, but it's not easily possible without some fundamental changes in AeroSpace.

If we combine your and `@raisjn`'s approaches (in case there is already a refresh session running, you cancel the newly submitted session **and** schedule another session in 100ms) that would be the solution that I'm happy to merge.

But please keep in mind that I'm currently in progress of [more fundamental refactoring](https://github.com/nikitabobko/AeroSpace/issues/131#issuecomment-2764689528) which will cover your and `@raisjn`'s suggestions anyway.

You are welcome to submit the PR with the mention corrections. If it takes me too long to do the fundamental refactoring, I will make an intermediate release with the discussed fix

**Reply by rickyz** [reply]
Oooh, excited to hear about the fixes in https://github.com/nikitabobko/AeroSpace/issues/131!

Re your comment, I don't quite follow the bit about cancellation - https://github.com/rickyz/AeroSpace/commit/a973bc4158834c4887712bdd82ba42b99b76c56c doesn't introduce any parallelism, so at the moment `scheduleRefreshAndLayout` is called (which is always from the main thread), there is never any currently running request.

What this is optimizing is the case where we have queued up the following work on the main thread while it was busy:

[AX event 1] [AX event 2] ... [AX event N]

before my change, processing of each AX event runs `refreshAndLayout()` inline. With my change, the N AX events are processed without performing any calls to `refreshAndLayout()`. After processing the events 1 .. N (which should be fast,  since event processing no longer calls `refreshAndLayout()`), the main thread will run `refreshAndLayout()` once.

**Reply by nikitabobko** [reply]
Aaah, thanks for the elaboration, I misread your commit. Please submit a PR, I'm happy to merge the commit as is.

**Comment by nikitabobko** [comment, marked as answer]
0.18.0 features big performance improvements https://github.com/nikitabobko/AeroSpace/releases/tag/v0.18.0-Beta https://github.com/nikitabobko/AeroSpace/issues/131. The published release covers the fixes in `@raisjn` fork.

I'm now all ears

**Comment by voidpointer0** [comment]
Thank you to everyone who contributed to this release!

I can finally use Aerospace while working (Unity). I experience minor lags when switching between workspaces when Unity is on, but previously it made the system completely unusable, and now it's barely noticeable. I will leave it open for a couple of hours and edit my comment if needed, but so far, an hour in, this release makes a giant difference.

Aerospace version: 0.18.0-Beta
Sequoia 15.3.1
Apple M1 Max 64GB RAM

**Comment by vianneychin** [comment]
Started using the update today and wow it is so fast and all my lag issues have disappeared. Switching is lightning fast.  🙌
